packages/cli/generation/ir-generator/src/utils/getCyclicTypes.ts

```diff
@@ -0,0 +1,203 @@
+        definitionFile: args.definitionFile,
+        rootApiFile: args.workspace.definition.rootApiFile.contents,
+        casingsGenerator: CASINGS_GENERATOR
+    });
+    const resolvedType: ResolvedType | undefined = args.typeResolver.resolveType({
+        type: args.typeName,
+        file: fileContext
+    });
+    if (resolvedType == null) {
+        return null;
+    }
+    if (resolvedType._type !== "named") {
+        return null;
+    }
+
+    // If this is a union type, every object path must be cyclic in order for the type to be considered an invalid cyclic type. This is also why we ignore container types, since they can be empty.
+    const isUnion =
+        isRawDiscriminatedUnionDefinition(resolvedType.declaration) ||
+        isRawUndiscriminatedUnionDefinition(resolvedType.declaration);
+
+    const typeProperties: TypeOrProperty[] = isUnion
+        ? unboxUnionTypes(resolvedType, fileContext, args.typeResolver)
+        : unboxObjectType(args, resolvedType);
+
+    return isUnion
+        ? checkUnionChildTypes(args, typeProperties.flatMap(getTypeParams))
+        : checkObjectChildTypes(args, typeProperties.flatMap(getTypeParams));
+}
+
+function unboxObjectType(args: CyclicTypeArgs, t: ResolvedType.Named): TypeOrProperty[] {
+    if (isRawObjectDefinition(t.declaration)) {
+        return getAllPropertiesForType(args) as TypeOrProperty[];
+    }
+    return [];
+}
+
+function unboxUnionTypes(
+    union: ResolvedType.Named,
+    context: FernFileContext,
+    typeResolver: TypeResolver
+): TypeOrProperty[] {
+    // helper for extracting typenames from union values
+    const unboxUnionValue = (value: UnionValue) => (typeof value === "string" ? value : value.type);
+    // helper for getting resolved types from union values
+    const getUnionResolvedTypes = (unionValue: UnionValue): TypeOrProperty[] => {
+        const value = unboxUnionValue(unionValue);
+        if (typeof value === "string") {
+            const resolvedType = typeResolver.resolveType({ type: value, file: context });
+            if (resolvedType != null) {
+                return [{ keys: ["union"], type: resolvedType }];
+            }
+        }
+        return [{ keys: ["union"], type: null }];
+    };
+    if (isRawDiscriminatedUnionDefinition(union.declaration)) {
+        const declaration = union.declaration as RawSchemas.DiscriminatedUnionSchema;
+        const u = declaration.union;
+        const values = Object.values(u) as UnionValue[];
+        return values.flatMap(getUnionResolvedTypes);
+    } else if (isRawUndiscriminatedUnionDefinition(union.declaration)) {
+        const declaration = union.declaration as RawSchemas.UndiscriminatedUnionSchema;
+        return declaration.union.flatMap(getUnionResolvedTypes);
+    }
+    return [];
+}
+
+function getTypeParams(t: TypeOrProperty): TypeParams | null {
+    const isObjectProperty = (t as ObjectPropertyWithPath).propertyType !== undefined;
```

Comment:
nit: would prefer a type guard instead of doing the cast twice

Reply:
Makes sense, I'll clean that up on the next version.